spark3/src/test/java/org/apache/iceberg/spark/source/TestStructuredStreamingRead3.java

```diff
@@ -0,0 +1,580 @@
+        .mode("append")
+        .save(location.toString());
+
+    df = spark.createDataFrame(avroFileRecords, SimpleRecord.class);
+    df.select("id", "data").write()
+        .format("iceberg")
+        .option("write-format", "avro")
+        .mode("append")
+        .save(location.toString());
+
+    table.refresh();
+    List<SimpleRecord> actual;
+
+    try {
+      Dataset<Row> ds = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+      StreamingQuery streamingQuery = ds.writeStream()
+          .format("memory")
+          .queryName("test12")
+          .outputMode(OutputMode.Append())
+          .start();
+      streamingQuery.processAllAvailable();
+      actual = spark.sql("select * from test12")
+          .as(Encoders.bean(SimpleRecord.class))
+          .collectAsList();
+
+      Assert.assertEquals(Stream.concat(Stream.concat(parquetFileRecords.stream(), orcFileRecords.stream()),
+          avroFileRecords.stream())
+          .collect(Collectors.toList()),
+          actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testReadStreamFromEmptyTable() throws IOException, TimeoutException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+
+    table.refresh();
+    List<SimpleRecord> actual;
+
+    try {
+      Dataset<Row> df = spark.readStream()
+          .format("iceberg")
+          .load(location.toString());
+      StreamingQuery streamingQuery = df.writeStream()
+          .format("memory")
+          .queryName("testemptytable")
+          .outputMode(OutputMode.Append())
+          .start();
+      streamingQuery.processAllAvailable();
+      actual = spark.sql("select * from testemptytable")
+          .as(Encoders.bean(SimpleRecord.class))
+          .collectAsList();
+
+      Assert.assertEquals(Collections.emptyList(), actual);
+    } finally {
+      for (StreamingQuery query : spark.streams().active()) {
+        query.stop();
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  public void testReadStreamWithSnapshotTypeOverwriteErrorsOut() throws IOException, TimeoutException {
+    File parent = temp.newFolder("parent");
+    File location = new File(parent, "test-table");
+
+    HadoopTables tables = new HadoopTables(CONF);
+    PartitionSpec spec = PartitionSpec.builderFor(SCHEMA).bucket("id", 3).build();
+    Table table = tables.create(SCHEMA, spec, location.toString());
+    TableOperations ops = ((BaseTable) table).operations();
+    TableMetadata meta = ops.current();
+    ops.commit(meta, meta.upgradeToFormatVersion(2));
+
```

Comment:
It would be very helpful to me if these tests extracted some of their common code into helper functions so it was easier to determine what the tests are doing.

Reply:
did another round of refactoring on the unittest code. 
pl. see if that helped.

Reply:
@RussellSpitzer - am trying to resolve the conversations whichever I addressed / incorporated your feedback. pl. feel free to re-open or point on specific places again if you feel the need.